plutus-core/plutus-core/src/Data/SatInt.hs

```diff
@@ -0,0 +1,297 @@
+ | otherwise = -- Common case: x1 >= x2 >= y
+               let !delta = x2 -# x1 -- <= 0
+                   !y' = y -# delta  -- y <= y' <= x1; hence y' is representable
+
+                   -- Invariant: x >= y
+                   -- Note that: z >= y' => z + delta won't underflow
+                   -- so we are guaranteed not to underflow if/when we recurse
+                   go_dn x | isTrue# (x <# y') = [SI (I# x)]
+                           | otherwise = SI (I# x) : go_dn (x +# delta)
+   in SI (I# x1) : go_dn x2
+
+
+-- Requires x2 <= x1
+efdtIntDnFB :: (SatInt -> r -> r) -> r -> Int# -> Int# -> Int# -> r
+efdtIntDnFB c n x1 x2 y    -- Be careful about underflow!
+ | isTrue# (y ># x2) = if isTrue# (y ># x1) then n else SI (I# x1) `c` n
+ | otherwise = -- Common case: x1 >= x2 >= y
+               let !delta = x2 -# x1 -- <= 0
+                   !y' = y -# delta  -- y <= y' <= x1; hence y' is representable
+
+                   -- Invariant: x >= y
+                   -- Note that: z >= y' => z + delta won't underflow
+                   -- so we are guaranteed not to underflow if/when we recurse
+                   go_dn x | isTrue# (x <# y') = SI (I# x) `c` n
+                           | otherwise = SI (I# x) `c` go_dn (x +# delta)
+               in SI (I# x1) `c` go_dn x2
+
+-- The following code is copied/adapted from GHC.Real.
+
+instance Real SatInt where
+  toRational (SI x) = toInteger x % 1
+
+instance Integral SatInt where
+    toInteger (SI (I# i)) = smallInteger i
+
+    SI a `quot` SI b
+     | b == 0                     = divZeroError
+     | a == minBound && b == (-1) = minBound
+     | otherwise                  = SI (a `quotInt` b)
+
+    SI a `rem` SI b
+     | b == 0                     = divZeroError
+     | a == minBound && b == (-1) = minBound
+     | otherwise                  = SI (a `remInt` b)
+
+    SI a `div` SI b
+     | b == 0                     = divZeroError
+     | a == minBound && b == (-1) = minBound
+     | otherwise                  = SI (a `divInt` b)
+
+    SI a `mod` SI b
+     | b == 0                     = divZeroError
+     | a == minBound && b == (-1) = minBound
+     | otherwise                  = SI (a `modInt` b)
+
+    SI a `quotRem` SI b
+     | b == 0                     = divZeroError
+     | a == minBound && b == (-1) = minBound
+     | otherwise                  =  a `quotRemSI` b
+
+    SI a `divMod` SI b
+     | b == 0                     = divZeroError
+     | a == minBound && b == (-1) = minBound
+     | otherwise                  =  a `divModSI` b
+
+quotRemSI :: Int -> Int -> (SatInt, SatInt)
+quotRemSI a@(I# _) b@(I# _) = (SI (a `quotInt` b), SI (a `remInt` b))
+
+divModSI ::  Int -> Int -> (SatInt, SatInt)
+divModSI x@(I# _) y@(I# _) = (SI (x `divInt` y), SI (x `modInt` y))
+
+plusSI :: SatInt -> SatInt -> SatInt
+plusSI (SI (I# x#)) (SI (I# y#)) =
+  case addIntC# x# y# of
+    (# r#, _ #) -> SI (I# r#)
+
+minusSI :: SatInt -> SatInt -> SatInt
```

Comment:
I wonder if we should remove things like minus and div, although that would of course mess up the `Num` instance and so on.  I think we really only need addition and multiplication (although there might be somewhere where I divide by 1000 or something: if so, I hope that'll go away soon). 

The reason I'm thinking this is that if we have this library lying about then there might be a temptation to use it for other things, and that could be dangerous if you're not careful.  You might have `x+y-z` where the addition 'overflows' to the maximum value and then the subtraction turns that back into something which looks reasonable but is in fact incorrect.  It might be best not to take that risk, or at least to have a `LOUD COMMENT` warning prospective users.

Reply:
I guess. It's pretty clear that it does saturating operations - if that's not what you want you shouldn't use this library!